Improve crash handling

dkasak opened the pull request:
This PR improves handling of crashing test cases in the corpus in three ways:
  *  Sets `AFL_CMIN_ALLOW_ANY` only during the initial test run, improving compatibility with the original afl-cmin. It also enables the `-C` flag to work as intended.
  * Detects and reports number of crashes encountered during corpus minimization.
  * Adds a `-c <dir>` flag to copy detected crashes into `<dir>`.

dkasak committed:
Set AFL_CMIN_ALLOW_ANY only for the first run, like afl-cmin.

Always setting AFL_CMIN_ALLOW_ANY includes crashes in the final output
of the cmin operation. This is unlike the original afl-cmin tool, which
only uses it for the first test run. It also prevents the -C flag from
functioning properly.

dkasak committed:
Implement ability to filter out crashes into a separate directory.

Crashes are always deduplicated by hash (SHA1), even if --no-dedup is
passed.

dkasak committed:
Mention -c in README.

dkasak commented:
Ping? Any feedback on this?

kcwu commented:
Thanks for reminding. Sorry I forgot this pull request.

kcwu commented on afl-cmin.py:
```diff
@@ -312,6 +329,22 @@ def main():
+                try:
+                    shutil.copy(crash_path, output_path)
+                except shutil.Error:
+                    # files are the same; do nothing since we already have it
```
I don't understand this case. They are dedup'ed, how are they still "the same"?

dkasak replied:
I think this can happen when files are hardlinks of one another (i.e. when they have the same inode). This happened to me several times, so I thought it be best to handle this edge case.

dkasak replied:
I confirm that this happens when the src and dest are hardlinks of one another. I clarified the comment to reflect this.

kcwu commented on afl-cmin.py:
```diff
@@ -312,6 +329,22 @@ def main():
             already_have.add(t)
         count += 1
 
+    if args.crash_dir:
```
Before your patch, crashes will be stored in `output_path`.
But after your patch, crashes will be skip if `crash_dir` is not specified.
Could you preserve the previous behavior?

dkasak replied:
Are you sure about this? I think this is only true if `AFL_CMIN_ALLOW_ANY=1` is set since otherwise `afl-showmap` doesn't return coverage for a crash so they are never included in the minimized set.

In any case, I've made afl-cmin.py have the same behaviour as before if `--crash-dir` is not passed. I did a quick test to ensure it works, namely:

1. `afl-cmin.py -i in -o out ...` doesn't retain crashes in `out` and doesn't save them anywhere.

   It also doesn't print how many crashes it encountered. I'm not too sure about this. Do we want it to always print the number of crashes?

2. `AFL_CMIN_ALLOW_ANY=1 afl-cmin.py -i in -o out ...` considers crashes as regular inputs and retains them if they happen to be effective (have novel tuples and less in size than all other files that have those tuples).
3. `afl-cmin.py -i in -o out --crash-dir crash ...` separates out crashes into the `crash` dir and prints how many it found when printing the number of effective files.

Anything else?

dkasak commented:
Err, so sorry, I forgot about this. I'll try to handle it soon.

dkasak committed:
Don't dedup crash files if we previously did a global dedup.

dkasak committed:
Use --crash-dir parameter name instead of -c.

dkasak committed:
Ensure that we only modify copies of the env.

dkasak committed:
Don't calculate crashed status inline.

dkasak committed:
Clarify shutil.Error comment.

dkasak committed:
If --crash-dir is not specified, treat crashes as ordinary inputs.

In other words, count them towards total coverage and preserve them as
part of the minimized corpus.

dkasak commented:
Finally got around to doing this.

kcwu commented on README.md:
```diff
@@ -7,6 +7,7 @@ Reimplement afl-cmin in python. Use less memory, less disk space, and faster.
  - Support the same command line flags as original afl-cmin.
  - dedup by hash of file content in the beginning.
  - `-i DIR` can be specified multiple times. Also support globbing.
+ - `-c DIR` to copy detected crashes (deduplicated by hash) into `DIR`.
```
--crash-dir

dkasak replied:
Done!

kcwu reviewed (changes requested):
Please update README.md, besides that, all looks good.

Regarding to the behavior without --cache-dir, I guess because AFL_CMIN_ALLOW_ANY=1 is always set by my original code. But I already forgot why I want to set it all the time. I have not used this script for almost two years.

dkasak committed:
Update README: `-c` -> `--crash-dir`
